File: languages/typescript/mem-large-json-transform/transform.mjs
#!/usr/bin/env node
import fs from 'node:fs';

const path = process.argv[2];
const data = JSON.parse(fs.readFileSync(path, 'utf8'));
const agg = new Map();
for (const item of data) {
  if (!agg.has(item.category)) {
    agg.set(item.category, { count: 0, value_sum: 0, weight_sum: 0, active_count: 0 });
  }
  const bucket = agg.get(item.category);
  bucket.count += 1;
  bucket.value_sum += item.value;
  bucket.weight_sum += item.weight;
  bucket.active_count += item.active ? 1 : 0;
}
const categories = Object.fromEntries(agg.entries());
console.log(JSON.stringify({ total_records: data.length, categories }));

File: languages/python/mem-large-json-transform/transform.py
#!/usr/bin/env python3
import json
import sys
from collections import defaultdict

path = sys.argv[1]
with open(path, 'r') as f:
    data = json.load(f)

agg = defaultdict(lambda: {"count": 0, "value_sum": 0, "weight_sum": 0, "active_count": 0})
for item in data:
    bucket = agg[item['category']]
    bucket['count'] += 1
    bucket['value_sum'] += item['value']
    bucket['weight_sum'] += item['weight']
    bucket['active_count'] += 1 if item['active'] else 0

summary = {
    "total_records": len(data),
    "categories": dict(agg),
}
print(json.dumps(summary, sort_keys=True))
